client/src/services/hotkeyManager.ts

```diff
@@ -0,0 +1,117 @@
+
+export enum Modifiers {
+  None = 1 << 0,
+  Ctrl = 1 << 1,
+  Alt = 1 << 2,
+  Shift = 1 << 3,
+}
+
+export enum Platforms {
+  Windows = 1 << 0,
+  MacOS = 1 << 1,
+  Linux = 1 << 2,
+  Web = 1 << 3,
+  Mobile = 1 << 4,
+  Desktop = Windows | MacOS | Linux,
+  All = Desktop | Web | Mobile,
+}
+
+export interface Hotkey {
+  key: string;
+  modifiers: number;
+  platforms: number;
+  command: () => Promise<void>;
+}
+
+class Hotkeys {
+  currentHotkeys: Hotkey[];
+  id: number;
+  enabled: boolean;
+
+  constructor(id: number) {
+    this.currentHotkeys = [];
+    this.id = id;
+    this.enabled = true;
+  }
+
+  async add(key: string, modifiers: number, platforms: number, command: () => Promise<void>): Promise<void> {
+    this.currentHotkeys.push({ key: key, modifiers: modifiers, platforms: platforms, command: command });
+  }
+
+  async enable(): Promise<void> {
+    this.enabled = true;
+  }
+
+  async disable(): Promise<void> {
+    this.enabled = false;
+  }
+}
+
+export class HotkeyManager {
+  currentHotkeys: Hotkeys[];
+  index: number;
+
+  constructor() {
+    this.currentHotkeys = [];
+    this.index = 0;
+    window.addEventListener('keydown', async (event: KeyboardEvent): Promise<void> => await this.onKeyPress(event, this.currentHotkeys));
+  }
+
+  newHotkeys = (): Hotkeys => {
+    const newKeys = new Hotkeys(this.index);
+    this.index++;
+    this.currentHotkeys.push(newKeys);
+    return newKeys;
+  };
+
+  async unregister(toRemove: Hotkeys): Promise<void> {
+    this.currentHotkeys = this.currentHotkeys.filter((hotkeys) => hotkeys.id !== toRemove.id);
+  }
+
+  private async onKeyPress(event: KeyboardEvent, hotKeys: Array<Hotkeys>): Promise<void> {
+    if (!isDesktop() && !isWeb()) {
+      return;
+    }
+    for (const keyPresses of hotKeys) {
+      for (const keyPress of keyPresses.currentHotkeys) {
+        if (this.checkPlatforms(keyPress.platforms)) {
+          continue;
+        }
+        if (event.key.toLowerCase() === keyPress.key) {
+          if (this.checkModifiers(event, keyPress.modifiers)) {
+            continue;
+          }
+          event.preventDefault();
+          if (keyPresses.enabled) {
+            await keyPress.command();
+          }
+        }
+      }
+    }
+  }
+
+  private checkModifiers(event: KeyboardEvent, modifiers: number): boolean {
+    // Returns true if needed modifiers aren't pressed or if not needed modifiers are pressed
+    const ctrlKey = isMacOS() ? event.metaKey : isDesktop() ? event.ctrlKey : event.ctrlKey || event.metaKey;
+    return ((modifiers & Modifiers.Ctrl && !ctrlKey) ||
+      (modifiers & Modifiers.Shift && !event.shiftKey) ||
+      (modifiers & Modifiers.Alt && !event.altKey) ||
+      (ctrlKey && !(modifiers & Modifiers.Ctrl)) ||
+      (event.shiftKey && !(modifiers & Modifiers.Shift)) ||
+      (event.altKey && !(modifiers & Modifiers.Alt))) as boolean;
+  }
+
+  private checkPlatforms(platforms: number): boolean {
```

Comment:
`checkPlatforms` and `checkModifiers` are weird names for functions returning booleans. In fact, I thought they were doing the opposite of what the comments say (return true if we're on the right platforms / if it has the right modifiers).